cocotb/share/include/gpi_logging.h

```diff
@@ -30,39 +30,183 @@
+ */
 enum gpi_log_levels {
-    GPIDebug = 10,
-    GPIInfo = 20,
-    GPIWarning = 30,
-    GPIError = 40,
-    GPICritical = 50
+    GPIDebug    = 10,   ///< Prints `DEBUG`    . Verbose information, useful for debugging
+    GPIInfo     = 20,   ///< Prints `INFO`     . Information about major events in the current program
+    GPIWarning  = 30,   ///< Prints `WARN`     . Encountered a recoverable bug, or information about surprising behavior
+    GPIError    = 40,   ///< Prints `ERROR`    . An unrecoverable error
+    GPICritical = 50    ///< Prints `CRITICAL` . An unrecoverable error, to be followed by immediate simulator shutdown
 };
 
+/** Logs a message at DEBUG log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+ */
+#define LOG_DEBUG(...) \
+    gpi_log("cocotb.gpi", GPIDebug,         __FILE__, __func__, __LINE__, __VA_ARGS__);
+
+/** Logs a message at INFO log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+ */
+#define LOG_INFO(...) \
+    gpi_log("cocotb.gpi", GPIInfo,          __FILE__, __func__, __LINE__, __VA_ARGS__);
+
+/** Logs a message at WARN log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+ */
+#define LOG_WARN(...) \
+    gpi_log("cocotb.gpi", GPIWarning,       __FILE__, __func__, __LINE__, __VA_ARGS__);
+
+/** Logs a message at ERROR log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+ */
+#define LOG_ERROR(...) \
+    gpi_log("cocotb.gpi", GPIError,         __FILE__, __func__, __LINE__, __VA_ARGS__);
+
+/** Logs a message at CRITICAL log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+ */
+#define LOG_CRITICAL(...) \
+    gpi_log("cocotb.gpi", GPICritical,      __FILE__, __func__, __LINE__, __VA_ARGS__);
+
+/** Logs a message at the given log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+    @param level The gpi_log_levels at this to log the message
+ */
+#define LOG_(level, ...) \
+    gpi_log("cocotb.gpi", level,      __FILE__, __func__, __LINE__, __VA_ARGS__);
+
+/** Type of a log handler function.
+    @param userdata  private implementation data registered with this function
+    @param name      Name of the logger
+    @param level     Level at which to log a message at
+    @param pathname  Name of the file where the call site is located
+    @param funcname  Name of the function where the call site is located
+    @param lineno    Line number of the call site
+    @param msg       The message to log, uses C-sprintf-style format specifier
+    @param ...       Additional arguments; formatted at inserted in message according to format specifier in msg argument
```

Comment:
```suggestion
    @param args      Additional arguments; formatted and inserted in message according to format specifier in msg argument
```

"formatted and inserted" in the other doc comments below as well.